### PR TITLE
Retrofit ArgumentFactory and NamedArgumentFinder with prepare-able interfaces, and fix some performance issues

### PR DESCRIPTION
**Pull request by stevenschlansker**
from
```
Benchmark                                  Mode  Cnt   Score   Error  Units
BeanBindingBenchmark.batchJdbc            thrpt    5  22.338 ± 2.203  ops/s
BeanBindingBenchmark.batchJdbiBean        thrpt    5   1.225 ± 0.097  ops/s
BeanBindingBenchmark.batchJdbiMap         thrpt    5   1.931 ± 0.244  ops/s
BeanBindingBenchmark.batchJdbiNamed       thrpt    5   1.933 ± 0.146  ops/s
BeanBindingBenchmark.batchJdbiPositional  thrpt    5   1.817 ± 0.178  ops/s

BeanBindingBenchmark.oneJdbc              thrpt    5  40.092 ± 2.037  ops/s
BeanBindingBenchmark.oneJdbi              thrpt    5   0.761 ± 0.030  ops/s
```

to
```
Benchmark                                  Mode  Cnt   Score   Error  Units
BeanBindingBenchmark.batchJdbc            thrpt    5  25.705 ± 3.563  ops/s
BeanBindingBenchmark.batchJdbiBean        thrpt    5  13.661 ± 1.808  ops/s
BeanBindingBenchmark.batchJdbiMap         thrpt    5   8.933 ± 1.956  ops/s
BeanBindingBenchmark.batchJdbiNamed       thrpt    5  11.945 ± 2.075  ops/s
BeanBindingBenchmark.batchJdbiPositional  thrpt    5  10.477 ± 1.130  ops/s

BeanBindingBenchmark.oneJdbc              thrpt    5  43.571 ± 3.520  ops/s
BeanBindingBenchmark.oneJdbi              thrpt    5  15.637 ± 1.170  ops/s
```

### CODE REVIEW

File: core/src/main/java/org/jdbi/v3/core/statement/BaseStatement.java [hunk 1]
```diff
@@ -32,15 +31,6 @@
     BaseStatement(Handle handle) {
         this.handle = handle;
         this.ctx = new StatementContext(handle.getConfig().createCopy(), handle.getExtensionMethod());
-
-        // Prevent bogus signatures like Update extends SqlStatement<Query>
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/core/statement/BaseStatement.java]: This check was about 15% of overall time to execute a single statement :O

Comment (qualidafial) [reply core/src/main/java/org/jdbi/v3/core/statement/BaseStatement.java]: Wow :grimacing:

**Comment by Alexey1Gavrilov** [comment]
I would like to see this PR merged. Do you need help with testing / reviewing this PR?

**Comment by stevenschlansker** [comment]
Yes please!  The code is rather dense, unfortunately, and the other `jdbi` contributors have had more pressing other concerns as of late and haven't had time to code review.  If you have time to review and offer thoughts, or test your own projects against a build of it, that'd be valuable.  It works for me but I'm hesitant to "ship it to production" for such a prominent library without at least some other people vouching for it! :smile_cat:

File: core/src/main/java/org/jdbi/v3/core/mapper/reflect/internal/BeanPropertiesFactory.java [hunk 2]
```diff
@@ -135,10 +98,33 @@ public T build() {
         static class BeanPojoProperty<T> implements PojoProperty<T> {
             final PropertyDescriptor descriptor;
             final QualifiedType<?> qualifiedType;
+            final ConcurrentMap<Class<?>, Optional<Annotation>> annoCache = new ConcurrentHashMap<>();
+            final Optional<Function<Object, Object>> getter;
```
Comment (Alexey1Gavrilov) [inline core/src/main/java/org/jdbi/v3/core/mapper/reflect/internal/BeanPropertiesFactory.java]: I'd recommend against using Optional<> as a field type, it is intended for return method type. as described here http://blog.joda.org/2014/11/optional-in-java-se-8.html

File: core/src/main/java/org/jdbi/v3/core/qualifier/QualifiedType.java [hunk 3]
```diff
@@ -40,6 +40,7 @@
 public final class QualifiedType<T> {
     private final Type type;
     private final Set<Annotation> qualifiers;
+    private int hashCode;
```
Comment (Alexey1Gavrilov) [inline core/src/main/java/org/jdbi/v3/core/qualifier/QualifiedType.java]: May be make it final and calculate in the constructor?

File: core/src/main/java/org/jdbi/v3/core/config/ConfigRegistry.java [hunk 4]
```diff
@@ -30,7 +35,11 @@
      * Creates a new config registry.
      */
     public ConfigRegistry() {
-        configs.put(JdbiCaches.class, new JdbiCaches());
+        get(JdbiCaches.class);
```
Comment (Alexey1Gavrilov) [inline core/src/main/java/org/jdbi/v3/core/config/ConfigRegistry.java]: I'd avoid if possible calling a non-private method from a constructor of a non-final class

File: core/src/main/java/org/jdbi/v3/core/qualifier/QualifiedType.java (hunk 3 above)
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/qualifier/QualifiedType.java]: I'm going to leave this one as is: making this change seems to regress the benchmarks.  I updated the code to better follow the pattern in String for lazily-computed hashes, assuming the JDK authors know better than I do :)

**Pull request merged**